### ISSUE TITLE
Zim Hangs before Completing Paste and Undo

### ISSUE THREAD

**Issue by LazyJ** [zim-desktop-wiki/zim-desktop-wiki#2409]
**Describe the bug**
Pasting from the clipboard or using the Undo function causes Zim to become unresponsive and unusable for several seconds until the pasted text appears or the Undo function reverts the recently changed text.

**To Reproduce**
1. For Paste problem: Copy or Cut some text.
2. Paste that text somewhere else in the document.
3. Wait for Zim to become responsive again.
4. For Undo problem:  delete a small section of text
5. Press Ctrl+Z
6. Wait for Zim to become responsive again.

**Expected behavior**
Pasted or reverted text to appear almost instantly.

**Screenshots**
N/A

**Environment (please complete the following information):**
 - Version: Zim 0.75.2
 - OS: Linux Mint 21.1
 - Language: English - US

**Additional context**
This bug just appeared in Zim v0.75.2, didn't occur in previous version.

For issues triggered by specific page content, please add an (anonymized) example that allows reproducing this issue.

For issues with no visible output / error, check the debug log for additional logging. This is available via the menu or by running "zim -D" from the commandline.

**Comment by jaap-karssenberg** [comment]
Sorry, but can't reproduce (Ubuntu / Gnome environment) - any specifics you can add that might help diagnose, like the debug log ?

**Comment by LazyJ** [comment]
> Sorry, but can't reproduce (Ubuntu / Gnome environment) - any specifics you can add that might help diagnose, like the debug log ?

Here's the snippet that appeared in the log when I Pasted two words:

`DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree`

`DEBUG: Requesting data for ('text/x-zim-parsetree', 0, 1) -- using set_with_data() workaround`


When I did an Undo:


`DEBUG: Action: undo`

`DEBUG: Action: redo`

**Comment by LazyJ** [comment]
Some new info to add:

So far there have been two updates to Zim today.

This *morning* the Zim update in the Update Manger was version:
`0.75.2-t202307061816~ubuntu22.04.1`

After installing the update is when the delayed Paste problem appeared.

This *afternoon* a second update was available:
`0.75.2-t202307071235~ubuntu22.04.1`

After installing the second update the delayed Paste problem persists.

`DEBUG: Targets available for paste: [Gdk.Atom.intern("TIMESTAMP", False), Gdk.Atom.intern("TARGETS", False), Gdk.Atom.intern("MULTIPLE", False), Gdk.Atom.intern("SAVE_TARGETS", False), Gdk.Atom.intern("UTF8_STRING", False), Gdk.Atom.intern("COMPOUND_TEXT", False), Gdk.Atom.intern("TEXT", False), Gdk.Atom.intern("STRING", False), Gdk.Atom.intern("text/plain;charset=utf-8", False), Gdk.Atom.intern("text/plain", False)], we want parsetree`

`DEBUG: Requesting data for ('text/x-zim-parsetree', 0, 1) -- using set_with_data() workaround`

**Comment by max-momot** [comment]
Debian 12 Gnome - I don't see this bug.

**Comment by LazyJ** [comment]
If it helps, Linux Mint is based on Ubuntu and its default desktop environment is a derivative of Gnome:
`Cinnamon 5.6.8 tk: GTK 3.24.33 wm: muffin`


Could the paste delay problem be due to file size and indexing the change?

As a test to eliminate possibilities, I tried copy/pasting text between a couple plain-text editors -- no problem.

Then I copied text from each of those two other editors *(Mousepad and Xed - Gedit-clone that is Linux Mint's default text editor)* into a Zim note page -- the paste process lagged for a few seconds.

I wondered if the problem stemmed from the size of the Zim note page's text file *(614kb)*. I created a new, empty Zim note page and the paste process was instantaneous.  !?

I went back to the large Zim note page and tried the paste again. This time there was less delay but still a delay.

Next, I went to the menu bar --> Tools --> and clicked on "Check and Update Index" and clicked on it again when that previous process had completed.

Then I tried a small copy/paste within that large file and the paste process was instantaneous.

After that I opened every Zim notebook and ran the "Check and Update Index" on each one twice.

Then I tried a two-word copy/paste again and the problem returned.

Then I tried the same two-word copy/paste in a *small* Zim note page and the paste was instant.

Then I tried the same two-word copy/paste back into the *large* Zim note page and the paste was instant.

So the problem started after yesterday's update *(2023_07_07)* to Zim; checking and updating the index didn't help but copy/pasting into a small Zim note page temporarily fixed the problem.

The debug is the same as what I've posted earlier.

Will try the flush and rebuild index  via command line next.

If the issue *is* because of the file size, what is the safe limit?

**Comment by max-momot** [comment]
@LazyJ  
Spell-check enabled?

**Comment by LazyJ** [comment]
> @LazyJ Spell-check enabled?

@max-momot 
Yes, spellcheck is enabled. I can see where that could cause drag on large files.

107,394 words in the file I've been working with. Besides the shear volume to check, peppered about the file are non-standard nouns, jargon, and my own abbreviations -- a workload that most likely gives the spellcheck fits.

These writings are for me only, not for any professional purpose, so I could do without spellcheck.

The in-document search _(as opposed to the notebook-wide search)_ is why I keep such large, contiguous, "flat" files; less loading and skipping through multiple pages. But, as the saying goes: "form follows function" or in this case "function before form"; I can change from my habitual flat structure and break documents into smaller files if need be.

Thanks for the spellcheck suggestion.

**Comment by max-momot** [comment]
I experimented with large texts in ZIM and gave up on the idea. Python is too slow. So I switched to something similar to Zettelkasten method.

**Comment by LazyJ** [comment]
Having too many characters in a single document seems to be the root of the problem.

I broke my current project into text files that are around 300kb to 360kb in file size _(roughly 277,000 characters each)_.

Turning off spellcheck also helped.

**Comment by LazyJ** [comment]
Seems the problem was caused by having too many characters in a single document. Breaking large documents into smaller ones _(roughly 277,000 each or 300kb to 360kb file size)_ resolved the problem.

**Comment by vossenjp** [comment]
On 7/8/23 03:47 PM, LazyJ wrote:
> Having too many characters in a single document seems to be the root of the problem.

This is also a problem for me and for much the same reason: in-document scan/ref/search.

> I broke my current project into text files that are around 300kb to 360kb in file size /(roughly 277,000 characters each)/.
> 
> Turning off spellcheck also helped.
For spell check speed, see also: https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1497#issuecomment-1358082581

Later,
JP
--  -------------------------------------------------------------------
JP Vossen, CISSP | http://www.jpsdomain.org/ | http://bashcookbook.com/

**Comment by LazyJ** [comment]
@vossenjp

Thanks for the gtkspell info and link. I could never figure out why Zim wouldn't recognize gtkspell but would detect gtkspellchecker.

I followed the instructions in your link to switch to gtkspell and added the info to my Program Setup note page for Zim.


@jaap-karssenberg 

I forgot to thank you for creating Zim. Over the years I've tried various note programs but Zim hits the spot just right.
